runtime/src/bank_client.rs

```diff
@@ -136,4 +175,24 @@ mod tests {
         bank_client.send_message(&doe_keypairs, message).unwrap();
         assert_eq!(bank_client.get_balance(&bob_pubkey).unwrap(), 42);
     }
+
+    fn do_work<F, T>(client_ctor: F)
+    where
+        F: Fn() -> T,
+        F: 'static + std::marker::Sync + std::marker::Send,
+        T: Client,
+    {
+        let client = client_ctor();
+        println!("tx count: {:?}", client.get_transaction_count());
+    }
+
+    #[test]
+    fn test_bank_client_pass_client_in_closure() {
+        solana_logger::setup();
+        let (genesis_block, root_keypair) = GenesisBlock::new(100_000_000_000_000);
+        let bank = Bank::new(&genesis_block);
+        let bank_client = BankClient::new(bank);
+        let client_ctor = move || -> BankClient { bank_client.clone() };
+        do_work(client_ctor);
```

Comment:
@greg Could use some help with this.  I want to be able to pass the closure between threads but rust complains that the channel sender cannot be shared.  Agreed but in this case, it is not directly accessible by multiple threads and instead, a clone of it is which is safe.  Rust still complains with the following error, any suggestions?

```bash
error[E0277]: `std::sync::mpsc::Sender<solana_sdk::transaction::Transaction>` cannot be shared between threads safely
   --> runtime/src/bank_client.rs:196:9
    |
196 |         do_work(client_ctor);
    |         ^^^^^^^ `std::sync::mpsc::Sender<solana_sdk::transaction::Transaction>` cannot be shared between threads safely
    |
    = help: within `[closure@runtime/src/bank_client.rs:195:27: 195:72 bank_client:bank_client::BankClient]`, the trait `std::marker::Sync` is not implemented for `std::sync::mpsc::Sender<solana_sdk::transaction::Transaction>`
    = note: required because it appears within the type `bank_client::BankClient`
    = note: required because it appears within the type `[closure@runtime/src/bank_client.rs:195:27: 195:72 bank_client:bank_client::BankClient]`
note: required by `bank_client::tests::do_work`
   --> runtime/src/bank_client.rs:179:5
    |
179 | /     fn do_work<F, T>(client_ctor: F)
180 | |     where
181 | |         F: Fn() -> T,
182 | |         F: 'static + std::marker::Sync + std::marker::Send,
...   |
186 | |         println!("tx count: {:?}", client.get_transaction_count());
187 | |     }
    | |_____^
```

Reply:
@jackcmay, I can get back to my PR first thing tomorrow. Let me know if you make progress here. At a glance it looks like you need to declare a lifetime `do_work<'a, F, T>` and use `'a` instead of `'static`.  Also, I'm curious if you can avoid this whole pattern of passing in a closure.  I'll have to play with it.

Reply:
@garious The changes in this PR are working great for ThinClient (local and real cluster), but when I try BankClient I see that error.  Seems like it should work but Rust is not being smart about what is actually available via the closure.  I'd be happy to get rid of the closure if there is another clean way to pass a bunch of `Client`s.  Also, notice that I changed how the BankClient loop works.